ml/src/main/scala/org/apache/spark/ml/h2o/param/H2OTargetEncoderParams.scala

```diff
@@ -0,0 +1,125 @@
+    require(getInputCols() != null && getInputCols().nonEmpty, "The list of input columns can't be null or empty!")
+    val fields = flatSchema.fields
+    val fieldNames = fields.map(_.name)
+    require(fieldNames.contains(getLabelCol()),
+      s"The specified label column '${getLabelCol()}' was not found in the input dataset!")
+    for(inputCol <- getInputCols()) {
+      require(fieldNames.contains(inputCol),
+        s"The specified input column '$inputCol' was not found in the input dataset!")
+    }
+    val ioIntersection = getInputCols().intersect(getOutputCols())
+    require(ioIntersection.isEmpty,
+      s"""The columns [${ioIntersection.map(i => s"'$i'").mkString(", ")}] are specified
+         |as input columns and also as output columns. There can't be an overlap.""".stripMargin)
+    StructType(flatSchema.fields ++ getOutputCols().map(StructField(_, DoubleType, nullable = true)))
+  }
+
+  //
+  // List of Parameters
+  //
+  protected final val foldCol = new NullableStringParam(this, "foldCol", "Fold column name")
+  protected final val labelCol = new Param[String](this, "labelCol", "Label column name")
+  protected final val inputCols = new StringArrayParam(this, "inputCols", "Names of columns that will be transformed")
+  protected final val holdoutStrategy = new H2OTargetEncoderHoldoutStrategyParam(this,
+    "holdoutStrategy",
+    """A strategy deciding what records will be excluded when calculating the target average on the training dataset.
+      |Options:
+      | None        - All rows are considered for the calculation
+      | LeaveOneOut - All rows except the row the colculation is made for
+      | KMeans      - Only out-of-fold data is considered (The option requires foldCol to be set.)
+    """.stripMargin)
+  protected final val blending = new CaseClassParam[H2OTargetEncoderBlendingSettings](this,
+    "blending",
+    """If set, the target average becomes a weighted average of the group target value and the global target value of a given row.
+      |The weight is determined by the size of the given group that the row belongs to.
+      |Attributes:
+      | InflectionPoint - The bigger number it's, the bigger groups will consider the global target value as a component in the weighted average.
+      | Smoothing       - Controls the rate of transition between group target values and global target values.""".stripMargin)
+  protected final val noise = new CaseClassParam[H2OTargetEncoderNoiseSettings](this,
+    "noise",
+    """The settings affecting how much of noise will be added to the target average.
+      |Attributes:
+      | Amount - amount of random noise
+      | Seed   - a seed of the generator producing the random noise
+    """.stripMargin)
+
+  //
+  // Default values
+  //
+  setDefault(
+    foldCol -> null,
+    labelCol -> "label",
+    inputCols -> Array[String](),
+    holdoutStrategy -> H2OTargetEncoderHoldoutStrategy.None,
+    blending -> null,
```

Comment:
Blending and noise parameters are grouped together. if `blending` or `noise` is disabled, the particular feature is disabled. @jakubhava, I would like to know your opinion on this.

Reply:
Good question! Does it semantically make sense to group this feature under one new, specific configuration object?

Reply:
if you started to use this estimator and you would see parameters like `smoothing`, `inflectionPoint` would you implicitly understand that they are parameters of blending? If blending is disabled they don't have any impact on anything. We can pose a similar question about the `seed` parameter for `noise`.

An alternative could be to name the parameters with some prefix that groups them, but I'm not sure what is better.

Reply:
On the other hand, when I thing about it the users setting these properties probably have to know they require bleeding to be enabled. Maybe we can keep it this is simple.

At the end, it is easier for the user to set the options directly compared to creating some additional configuration object.

I would keep it like it is now